share/p2p/peers/manager.go

```diff
@@ -0,0 +1,350 @@
+package peers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+
+	logging "github.com/ipfs/go-log/v2"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/celestiaorg/celestia-node/header"
+	libhead "github.com/celestiaorg/celestia-node/libs/header"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/availability/discovery"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexsub"
+)
+
+//TODO: address TODOs
+//TODO: add backoff
+//TODO: add debug logs
+//TODO: simplify tests
+// TODO: add GC for old / unwanted pools
+//TODO: metrics
+// - validation results rate[result_type, source:discovery/shrexsub]
+// - peer retrival rate[shrexsub/discovery]
+// - retrieval time hist
+// - validation time hist
+// - blacklisted peers count
+// - blacklistedHash count
+
+const (
+	gcInterval = time.Second * 30
+
+	ResultSuccess syncResult = iota
+	ResultFail
+	ResultPeerMisbehaved
+)
+
+var log = logging.Logger("shrex/peer-manager")
+
+// Manager keeps track of peers coming from shrex.Sub and from discovery
+type Manager struct {
+	lock sync.Mutex
+
+	// header subscription is necessary in order to validate the inbound eds hash
+	headerSub libhead.Subscriber[*header.ExtendedHeader]
+	shrexSub  *shrexsub.PubSub
+	disc      *discovery.Discovery
+	host      host.Host
+	connGater *conngater.BasicConnectionGater
+
+	// pools collecting peers from shrexSub
+	pools           map[string]*syncPool
+	poolSyncTimeout time.Duration
+	// fullNodes collects full nodes peer.ID found via discovery
+	fullNodes *pool
+
+	// peers that misbehaved
+	blacklistedPeers map[peer.ID]bool
```

Comment:
we can avoid this field by relying on ConnGater. Please check this [link](https://github.com/libp2p/go-libp2p/blob/635cc650cfb3bbbdb4bef6ae29f45e3c8966fb31/p2p/net/conngater/conngater.go#L277)

Reply:
done